### PR TITLE
Study prep

### PR DESCRIPTION
**Pull request by squirrelo**
Requires #1582 

Adds the prep template information and graphs to the study page.

Still needs the `Artifact(id).to_html()` #1576  and `PrepTemplate(id).last_artifact` #1573 to be completed.

### CODE REVIEW

**Comment by antgonza** [comment]
@squirrelo could you update ?

**Comment by squirrelo** [comment]
I will update when #1582 is merged, as it is already a complicated conflict to do that update.

**Comment by squirrelo** [comment]
Updated

**Comment by antgonza** [comment]
Could you pull from the branch? What's the plan for #1576 and #1573? Are they going to be implemented in this PR or in another one?

**Comment by squirrelo** [comment]
They will be implemented as a different PR. Right now there are placeholders where that information will go once it is available.

**Comment by antgonza** [comment]
OK, thanks. Do you know who is implementing them or are they part of an open PR?

**Comment by squirrelo** [comment]
They will be implemented by @josenavas, no current PR is in.

**Comment by josenavas** [comment]
@squirrelo see #1599 for how to get the youngest artifact in the lineage. Since that is something that has to do with the lineage of the artifact, I don't think it is correct to add it to the prep template, and I think it should be a call in the artifact. Thus, the call should look something like: `PrepTemplate(id).artifact.youngest_artifact`
Note that the name change is also to be consistent with the other nomenclature used in the Artifact object.

**Comment by squirrelo** [comment]
The `youngest_artifact` and `artifact.name` information are now implemented in this pull request. The to_html can wait and be added in as a micro-PR, so that should not block review of this PR.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 1]
```diff
@@ -6,52 +6,358 @@
...
+
+    Returns
+    -------
+    str
+        The investigation type chosen by the user
+    """
+    if inv_type == '':
+        inv_type = None
+    elif inv_type == 'Other' and user_def_type == 'New Type':
+        # This is a nre user defined investigation type so store it
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: what's nre?

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 2]
```diff
@@ -6,52 +6,358 @@
...
+def prep_template_post_req(study_id, user_id, prep_template, data_type,
+                           investigation_type=None,
+                           user_defined_investigation_type=None,
+                           new_investigation_type=None):
+    """Adds a prep template to the system
+
+    Parameters
+    ----------
+    study_id : int
+        Study to attach the pre ptemplate to
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: prep template.

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 1 above)
Comment (rob-knight) [inline qiita_pet/handlers/api_proxy/prep_template.py]: typo for “new”?

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 3]
```diff
@@ -6,52 +6,358 @@
...
+        Existing investigation type to attach to the prep template
+    user_defined_investigation_type: str, optional
+        Existing user added investigation type to attach to the prep template
+    new_investigation_type: str, optional
+        Investigation type to add to the system
     """
     access_error = check_access(study_id, user_id)
     if access_error:
         return access_error
-    # Can only pass ids over API, so need to instantiate object
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This block wasn't use before, right?

Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This code was moved to the study.py file in the api-proxy folder since it is a summary of what is in the study.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 4]
```diff
@@ -6,52 +6,358 @@
...
+    prep_template : str, optional
+        filepath to use for updating
+    investigation_type: str, optional
+        Existing investigation type to attach to the prep template
+    user_defined_investigation_type: str, optional
+        Existing user added investigation type to attach to the prep template
+    new_investigation_type: str, optional
+        Investigation type to add to the system
+    """
+    # Get the uploads folder
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This comment is wrong, right?

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 3 above)
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Cool, thanks for the clarification.

File: qiita_pet/handlers/api_proxy/artifact.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    -------
+    dict of lists of tuples
+        A dictionary containing the edge list representation of the graph,
+        and the node labels. Formatted as:
+        {'status': status,
+         'message': message,
+         'edge_list': [(0, 1), (0, 2)...],
+         'node_labels': [(0, 'label0'), (1, 'label1'), ...]}
+
+    Raises
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/artifact.py]: This per se doesn't raise anything, right?

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 6]
```diff
@@ -6,52 +6,358 @@
...
+    prep_template : str
+        Filepath to the prep template being added
+    data_type : str
+        Data type of the processed samples
+    investigation_type: str, optional
+        Existing investigation type to attach to the prep template
+    user_defined_investigation_type: str, optional
+        Existing user added investigation type to attach to the prep template
+    new_investigation_type: str, optional
+        Investigation type to add to the system
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This is missing returns

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 7]
```diff
@@ -6,52 +6,358 @@
...
+        The current user object id
+    prep_template : str, optional
+        filepath to use for updating
+    investigation_type: str, optional
+        Existing investigation type to attach to the prep template
+    user_defined_investigation_type: str, optional
+        Existing user added investigation type to attach to the prep template
+    new_investigation_type: str, optional
+        Investigation type to add to the system
+    """
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Missing returns.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 8]
```diff
@@ -6,52 +6,358 @@
...
+def prep_template_filepaths_get_req(prep_id, user_id):
+    """Returns all filepaths attached to a prep template
+
+    Parameters
+    ----------
+    prep_id : int
+        The current prep template id
+    user_id : int
+        The current user object id
+    """
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Missing returns

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 9]
```diff
@@ -6,52 +6,358 @@
...
+
+    Returns
+    -------
+    dict of lists of tuples
+        A dictionary containing the edge list representation of the graph,
+        and the node labels. Formatted as:
+        {'edge_list': [(0, 1), (0, 2)...],
+         'node_labels': [(0, 'label0'), (1, 'label1'), ...]}
+
+    Raises
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This doesn't raise anything, right?

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 10]
```diff
@@ -6,52 +6,358 @@
...
+            'User': sorted(ontology.user_defined_terms)
+            }
+
+
+def prep_template_graph_get_req(prep_id, user_id):
+    """Returns graph of all artifacts created from the prep base artifact
+
+    Parameters
+    ----------
+    study_id : int
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: No longer a parameter.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 11]
```diff
@@ -6,52 +6,358 @@
...
+from qiita_db.util import convert_to_id
 from qiita_db.study import Study
+from qiita_db.ontology import Ontology
+from qiita_db.metadata_template.prep_template import PrepTemplate
 
 
-def prep_template_get_req(study_id, user_id):
-    """Gives a summary of each prep template attached to the study
+@execute_as_transaction
+def _process_investigation_type(inv_type, user_def_type, new_type):
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This doesn't have an specific test, right?

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 12]
```diff
@@ -6,52 +6,358 @@
...
-            start_artifact = prep.artifact
-            info = {
-                'name': 'PREP %d NAME' % prep.id,
-                'id': prep.id,
-                'status': prep.status,
-                'start_artifact': start_artifact.artifact_type,
-                'start_artifact_id': start_artifact.id,
-                'last_artifact': 'TODO new gui'
+    fp_rpt = check_fp(study_id, prep_template)
+    if isinstance(fp_rpt, dict):
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This if is not tested

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 13]
```diff
@@ -6,52 +6,358 @@
...
+            data_type_id = convert_to_id(data_type, 'data_type')
+            # deleting previous uploads and inserting new one
+            PrepTemplate.create(load_template_to_dataframe(fp_rpt),
+                                Study(int(study_id)), int(data_type_id),
+                                investigation_type=investigation_type)
+            remove(fp_rpt)
+
+            # join all the warning messages into one. Note that this info
+            # will be ignored if an exception is raised
+            if warns:
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This is not tested.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 14]
```diff
@@ -6,52 +6,358 @@
...
+        and the values are list of tuples. Each tuple is an observed value in
+        the category and the number of times its seen.
+        Format {status: status,
+                message: message,
+                num_samples: value,
+                category: [(val1, count1), (val2, count2), ...], ...}
+    """
+    prep = PrepTemplate(int(prep_id))
+    access_error = check_access(prep.study_id, user_id)
+    if access_error:
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This is not tested.

Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Deleting, wrong line/if.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 15]
```diff
@@ -6,52 +6,358 @@
...
+        return access_error
+    df = prep.to_dataframe()
+    out = {'num_samples': df.shape[0],
+           'summary': {}}
+
+    # drop the prep_id column if it exists
+    if 'study_id' in df.columns:
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This is not tested.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 16]
```diff
@@ -6,52 +6,358 @@
...
+        Investigation type to add to the system
+    """
+    # Get the uploads folder
+    prep = PrepTemplate(int(prep_id))
+    study_id = prep.study_id
+    access_error = check_access(study_id, user_id)
+    if access_error:
+        return access_error
+
+    if investigation_type:
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Could you confirm that test_prep_template_put_req covers this section? Thanks.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 17]
```diff
@@ -6,52 +6,358 @@
...
+    -------
+    dict of list of str
+        {'ENA': [term1, term2, ...],
+         'User': [userterm1, userterm2, ...]}
+    """
+    # Get all the ENA terms for the investigation type
+    ontology = Ontology(convert_to_id('ENA', 'ontology'))
+    # make "Other" last on the list
+    ena_terms = sorted(ontology.terms)
+    ena_terms.remove('Other')
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Why do you remove it and then add it?

File: qiita_pet/handlers/api_proxy/util.py [hunk 18]
```diff
@@ -37,3 +40,31 @@ def check_access(study_id, user_id):
...
+    Parameters
+    ----------
+    study_id : int
+        Study file uploaded to
+    filename : str
+        name of the uploaded file
+
+    Returns
+    -------
+    dict or str
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/util.py]: To have a consistent return, should this always return a dict and if success return fp_rsp in file?

File: qiita_pet/handlers/api_proxy/util.py [hunk 19]
```diff
@@ -37,3 +40,31 @@ def check_access(study_id, user_id):
         return {'status': 'error',
                 'message': 'User does not have access to study'}
     return {}
+
+
+def check_fp(study_id, filename):
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/util.py]: No test for this function.

File: qiita_pet/handlers/study_handlers/prep_template.py [hunk 20]
```diff
@@ -0,0 +1,79 @@
+# -----------------------------------------------------------------------------
```
Comment (antgonza) [inline qiita_pet/handlers/study_handlers/prep_template.py]: One of the ideas of these changes was to create tests, right? However, there are no tests for this ...

**Comment by antgonza** [comment]
Done with first pass: tests missing and some minor comments. Also, looking at travis' errors it seems like most of them are solvable as some of the new additions are failing or are simple things. I know we agree that some failures are fine but I think the new additions shouldn't fail.

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 1 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/prep_template.py]: yup, fixed.

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 17 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/prep_template.py]: To make sure it is the last item in the list.

Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: OK. That means that is always there, right? Cause if not it's gonna fail. So what about always removing and adding in the template?

Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/prep_template.py]: `Other` is actually part of the ontology, so it is always there. Removing it completely from what is returned would alter the ontology, which we don't want to do.

Comment (antgonza) [inline qiita_pet/handlers/api_proxy/prep_template.py]: OK

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 16 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Added a specific test for here.

File: qiita_pet/handlers/api_proxy/util.py (hunk 18 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/util.py]: I suppose, although this means the contents of the `file` key will be different for error and non-error, since the full filepath will be there on success and just the filename on error.

Comment (antgonza) [inline qiita_pet/handlers/api_proxy/util.py]: Well, both of them are gonna return a string, right? The difference is what you do with that string.

**Comment by squirrelo** [comment]
Comments addressed.

**Comment by antgonza** [comment]
Could you minimize qiita_pet/static/vendor/css/vis.css? After that :+1:

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 21]
```diff
@@ -6,52 +6,383 @@
...
+        The investigation type chosen by the user
+    """
+    if inv_type == '':
+        inv_type = None
+    elif inv_type == 'Other' and user_def_type == 'New Type':
+        # This is a new user defined investigation type so store it
+        inv_type = new_type
+        ontology = Ontology(convert_to_id('ENA', 'ontology'))
+        ontology.add_user_defined_term(inv_type)
+    elif inv_type == 'Other' and user_def_type != 'New Type':
```
Comment (mortonjt) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This segment of code doesn't get tested (as far as I know).  Does this need a test?

File: qiita_pet/handlers/api_proxy/artifact.py [hunk 22]
```diff
@@ -0,0 +1,48 @@
+from qiita_db.artifact import Artifact
+from qiita_pet.handlers.api_proxy.util import check_access
+
+
+def artifact_graph_get_req(artifact_id, direction, user_id):
```
Comment (mortonjt) [inline qiita_pet/handlers/api_proxy/artifact.py]: This also doesn't have a test.  Are we ok with this?

**Comment by squirrelo** [comment]
tests added

File: qiita_pet/handlers/api_proxy/artifact.py [hunk 23]
```diff
@@ -0,0 +1,48 @@
+from qiita_db.artifact import Artifact
```
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/artifact.py]: This file is missing the Copyright information on top.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 24]
```diff
@@ -6,52 +6,383 @@
...
+                'message': message,
+                'num_samples': value,
+                'category': [(val1, count1), (val2, count2), ...], ...}
+    """
+    prep = PrepTemplate(int(prep_id))
+    access_error = check_access(prep.study_id, user_id)
+    if access_error:
+        return access_error
+    df = prep.to_dataframe()
+    out = {'num_samples': df.shape[0],
```
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Missing the 'status' and 'message' keywords.

Comment (josenavas) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Ooops - just saw that it is added later.

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 15 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/prep_template.py]: I don't think this type of handling should be done outside of the prep template. Why are you checking if the study_id is present? Isn't it showing a knowledge of how that data is stored outside of the DB?

Also, recent discussions showed that we may want to add the study_id as a column, since it will help for meta-analysis - but that is a different discussion.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 25]
```diff
@@ -6,52 +6,383 @@
...
+    # drop the prep_id column if it exists
+    if 'study_id' in df.columns:
+        df.drop('study_id', axis=1, inplace=True)
+    cols = list(df.columns)
+    for column in cols:
+        counts = df[column].value_counts()
+        out['summary'][str(column)] = [(str(key), counts[key])
+                                       for key in natsorted(counts.index)]
+    # Add expected messaging and status info
+    out.update({'status': 'success', 'message': ''})
```
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Why don't you add those keywords/values at the creation time of the dictionary? I don't see the usage of adding them now and it will make the code more readable.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 26]
```diff
@@ -6,52 +6,383 @@
...
+    investigation_type = _process_investigation_type(
+        investigation_type, user_defined_investigation_type,
+        new_investigation_type)
+
+    msg = ''
+    status = 'success'
+    try:
+        with warnings.catch_warnings(record=True) as warns:
+            # force all warnings to always be triggered
+            warnings.simplefilter("always")
```
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/prep_template.py]: I don't understand why you are forcing the warnings here - they should be already forced in the code base when the Warning is defined.

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 27]
```diff
@@ -6,52 +6,383 @@
...
+    access_error = check_access(prep.study_id, user_id)
+    if access_error:
+        return access_error
+    return {'status': 'success',
+            'message': '',
+            'filepaths': prep.get_filepaths()
+            }
+
+
+def prep_ontology_get_req():
```
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Why is the ontology information attached to the prep template?

File: qiita_pet/handlers/api_proxy/tests/test_prep_template.py [hunk 28]
```diff
@@ -6,28 +6,112 @@
...
+    def test_process_investigation_type_new_term(self):
+        randstr = ''.join([choice(ascii_letters) for x in range(30)])
+        obs = _process_investigation_type('Other', 'New Type', randstr)
+        self.assertEqual(obs, randstr)
+
+        # Make sure New Type added
+        ontology = Ontology(999999999)
+        self.assertIn(randstr, ontology.user_defined_terms)
+
+    def test_prep_ontology_get_req(self):
```
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/tests/test_prep_template.py]: Can we add a test in which the user defined terms are not empty?

File: qiita_pet/handlers/api_proxy/tests/test_prep_template.py [hunk 29]
```diff
@@ -6,28 +6,112 @@
...
+                       'Metagenomics', 'Pooled Clone Sequencing',
+                       'Population Genomics', 'RNASeq', 'Resequencing',
+                       'Synthetic Genomics', 'Transcriptome Analysis',
+                       'Whole Genome Sequencing', 'Other'],
+               'User': [],
+               'status': 'success',
+               'message': ''}
+        self.assertEqual(obs, exp)
+
     def test_prep_template_get_req(self):
```
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/tests/test_prep_template.py]: I think this test is incomplete. It is only testing that one of the samples is present. I'm not sure if that is extensive enough.

File: qiita_pet/handlers/api_proxy/tests/test_prep_template.py [hunk 30]
```diff
@@ -35,5 +119,184 @@ def test_prep_template_get_req_no_access(self):
                'message': 'User does not have access to study'}
         self.assertEqual(obs, exp)
 
+    def test_prep_template_post_req(self):
```
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/tests/test_prep_template.py]: It doesn't look like there is a test in which prep_template_post_req works correctly (i.e. adds a prep template to the system).

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 13 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/prep_template.py]: @squirrelo ping on this.

File: qiita_pet/handlers/api_proxy/tests/test_prep_template.py [hunk 31]
```diff
@@ -35,5 +119,184 @@ def test_prep_template_get_req_no_access(self):
...
+        self.assertEqual(obs, exp)
+
+    def test_prep_template_post_req_bad_filepath(self):
+        obs = prep_template_post_req(1, 'test@foo.bar', 'badfilepath', '16S')
+        exp = {'status': 'error',
+               'message': 'file does not exist',
+               'file': 'badfilepath'}
+        self.assertEqual(obs, exp)
+
+    def test_prep_template_put_req(self):
```
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/tests/test_prep_template.py]: This is also missing a test in which the function works with no issues.

File: qiita_pet/handlers/api_proxy/tests/test_prep_template.py [hunk 32]
```diff
@@ -35,5 +119,184 @@ def test_prep_template_get_req_no_access(self):
...
+        self.assertEqual(obs, exp)
+
+    def test_prep_template_put_req_bad_filepath(self):
+        obs = prep_template_put_req(1, 'test@foo.bar', 'badfilepath')
+        exp = {'status': 'error',
+               'message': 'file does not exist',
+               'file': 'badfilepath'}
+        self.assertEqual(obs, exp)
+
+    def test_prep_template_delete_req(self):
```
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/tests/test_prep_template.py]: There is no test in which the delete works.

File: qiita_pet/handlers/api_proxy/util.py (hunk 18 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/util.py]: This always returns a str, what is different is the contents of the 'file' key. @squirrelo please update

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 26 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This is so every warning is caught and recorded properly, so they can be displayed to the user.

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 27 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/prep_template.py]: Because this is the ENA ontology that is already attached to the prep template and prep template page currently.

File: qiita_pet/handlers/study_handlers/prep_template.py (hunk 20 above)
Comment (josenavas) [inline qiita_pet/handlers/study_handlers/prep_template.py]: @squirrelo I don't think the tests for this have been added.

One thing that I found useful when I'm addressing comments from a PR is to answer "done" to the comment, so I make sure that I don't forget to address any of the comments.

File: qiita_pet/handlers/study_handlers/prep_template.py [hunk 33]
```diff
@@ -0,0 +1,79 @@
...
+class PrepTemplateGraphAJAX(BaseHandler):
+    @authenticated
+    def get(self):
+        prep = to_int(self.get_argument('prep_id'))
+        self.write(prep_template_graph_get_req(prep, self.current_user.id))
+
+
+class PrepTemplateAJAX(BaseHandler):
+    @authenticated
+    def get(self):
```
Comment (josenavas) [inline qiita_pet/handlers/study_handlers/prep_template.py]: I still think there is too much work done in the handlers. Since we are already doing a rework of these handlers, I think it would be useful to start addressing #1145

Comment (squirrelo) [inline qiita_pet/handlers/study_handlers/prep_template.py]: I don't understand what you mean by too much work. All this is doing is requesting data to display in the HTML page. There is no processing of any information actually done here.

File: qiita_pet/handlers/study_handlers/sample_template.py [hunk 34]
```diff
@@ -39,10 +40,15 @@ def get(self):
 
     @authenticated
     def post(self):
-        """Edit/delete sample template"""
+        """Edit/delete/create sample template"""
```
Comment (josenavas) [inline qiita_pet/handlers/study_handlers/sample_template.py]: I'm not seeing delete in here?

File: qiita_pet/handlers/study_handlers/sample_template.py [hunk 35]
```diff
@@ -12,7 +12,8 @@
 from qiita_pet.handlers.util import download_link_or_path
 from qiita_db.util import get_files_from_uploads_folders
 from qiita_pet.handlers.api_proxy import (
-    sample_template_summary_get_req, sample_template_put_req,
+    sample_template_summary_get_req,
```
Comment (josenavas) [inline qiita_pet/handlers/study_handlers/sample_template.py]: I think this file is also missing tests?

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 26 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/prep_template.py]: I don't think this needs to be here. Note that below you have a similar function and you are not calling `warnings.simplefilter`. The reason for that is because the system is already set up so by default the warnings are always triggered. Please remove as it is not needed.

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 27 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/prep_template.py]: change name to `ena_ontology_get_req` as it is independent of the prep template.

File: qiita_pet/handlers/study_handlers/prep_template.py [hunk 36]
```diff
@@ -0,0 +1,79 @@
...
+        """Send formatted summary page of prep template"""
+        study_id = self.get_argument('study_id')
+        prep_id = self.get_argument('prep_id')
+        files = [f for _, f in get_files_from_uploads_folders(study_id)
```
Comment (josenavas) [inline qiita_pet/handlers/study_handlers/prep_template.py]: This should be done somewhere else

File: qiita_pet/handlers/study_handlers/prep_template.py (hunk 33 above)
Comment (josenavas) [inline qiita_pet/handlers/study_handlers/prep_template.py]: The problem is that you are not testing anywhere that the formatting that is done in this handlers is correct or not. What this means is that the interface can be broken in some cases without you noticing it because the tests are not exhaustive enough.

As I exposed in yesterday's meeting - ideally we want to be able to list all the possible ways to get to this handler and provide that as tests inputs. If that is done in a different function is way easier to test.

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 15 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This is code from the original implementation that was part of the study description handler, so I can not speak to the original intention of the code. I can remove this if wanted.

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 25 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/prep_template.py]: done

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 27 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/prep_template.py]: done

File: qiita_pet/handlers/api_proxy/tests/test_prep_template.py (hunk 28 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/tests/test_prep_template.py]: Done

File: qiita_pet/handlers/study_handlers/prep_template.py (hunk 20 above)
Comment (squirrelo) [inline qiita_pet/handlers/study_handlers/prep_template.py]: The tests exist in `qiita_pet/test/test_study_handlers.py`

File: qiita_pet/handlers/api_proxy/prep_template.py (hunk 13 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/prep_template.py]: test was added previously

File: qiita_pet/handlers/api_proxy/tests/test_prep_template.py (hunk 29 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/tests/test_prep_template.py]: Now tests all keys in template to make sure all samples are returned.

File: qiita_pet/handlers/api_proxy/util.py (hunk 18 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/util.py]: done

File: qiita_pet/handlers/study_handlers/sample_template.py (hunk 34 above)
Comment (squirrelo) [inline qiita_pet/handlers/study_handlers/sample_template.py]: line 55

File: qiita_pet/handlers/api_proxy/tests/test_prep_template.py (hunk 30 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/tests/test_prep_template.py]: Done

File: qiita_pet/handlers/api_proxy/tests/test_prep_template.py (hunk 31 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/tests/test_prep_template.py]: done

File: qiita_pet/handlers/api_proxy/tests/test_prep_template.py (hunk 32 above)
Comment (squirrelo) [inline qiita_pet/handlers/api_proxy/tests/test_prep_template.py]: done

File: qiita_pet/handlers/study_handlers/sample_template.py (hunk 35 above)
Comment (squirrelo) [inline qiita_pet/handlers/study_handlers/sample_template.py]: The tests exist in `qiita_pet/test/test_study_handlers.py`

**Comment by antgonza** [comment]
@squirrelo and @josenavas, what's the status of this PR?

File: qiita_pet/handlers/api_proxy/prep_template.py [hunk 37]
```diff
@@ -6,52 +6,379 @@
...
+           'summary': {},
+           'status': 'success',
+           'message': ''}
+
+    cols = list(df.columns)
+    for column in cols:
+        counts = df[column].value_counts()
+        out['summary'][str(column)] = [(str(key), counts[key])
+                                       for key in natsorted(counts.index)]
+    # Add expected messaging and status info
```
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/prep_template.py]: This comment is outdated.

**Comment by josenavas** [comment]
Just one comment @squirrelo - let me know when it is addressed and I will merge.

**Comment by squirrelo** [comment]
Done, ready to merge.

**Comment by antgonza** [comment]
Thanks!!
